students/Josh_HOff/lesson04/mailroom2.py

```diff
@@ -0,0 +1,79 @@
+donor = ''
+donation = 0
+tab = '    '
+
+#this function gets a donor's name and donation amount and adds them both to the donor list.
+def send_thank_you():
+    while True:
+        print('')
+        donor = input('Type a Full Name: ')
+        if switch_func_dict.get(donor, continue_func)() == False:
+            return
+        elif donor != 'list':
+            print('')
+            donation = input('What is the donation amount?: ')
+            if donation == 'quit':
+                return
+            donation = float(donation)
+            if donor in donors_list:
+                donors_list[donor] += [donation]
+            else:
+                donors_list[donor] = [donation]
+            print('')
+            print(f'Hello, {donor}! Thank you very much for your generous donation of ${donation:.2f}! Your contribution is essential and will be well utilized.')
+            
+#this function prints out a list of the donors in the donor list.            
+def print_list():
+    print('')
+    for i in donors_list:
+        print(i)
+    return True
+
+#this function prints out a report on the prompt screen.    
+def create_report():
+    print('')
+    y = '|'
+    print(f'Donor Name{y:>14} Total Given {y} Num Gifts {y} Average Gift')
+    print('-' * 63)
+    temp_set = donors_list.copy()
+    for i, val in donors_list.items():
+        temp_set[i] = sum(val)        
+    temp_set = dict(sorted(temp_set.items(), key=lambda kv: kv[1], reverse=True))
+    for i, val in temp_set.items():
+        gift = len(donors_list[i])
+        average = ( val / gift)
+        print(f'{i:<23} $ {val:>11.2f} {gift:>11} {average:>11.2f}')
+    print('')    
+
+#this function quits the previous menu.    
+def quitting():
+    return False
+
+#this function creates files for all donors in the donor list.
+def letters_to_everyone():
+#    pth = pathlib.Path('./donations/')
+#    p = pth.absolute()
+#    print(p)
+    for i, val in donors_list.items():
+        outfile = open(f'{i}.txt', 'w')
+        donation = sum(val)
+        outfile.write(f'Dear {i}, \n\n{tab}Thank you very much for your most recent donation \
+of ${val[-1]:.2f}! \n\n{tab}You have now donated a total of ${donation:.2f}. \n\n{tab}Your support \
+is essential to our success and will be well utilized. \n\n{tab*2}Sincerely, \n{tab*3}-The Company')
+        outfile.close()
+
+#this function will continue the previous while loop and prevents errors if a user enters an unexpected input.
+def continue_func():
+    return True
+    
+switch_func_dict = {'1':send_thank_you, '2':create_report, '3':letters_to_everyone, '4':quitting, 'quit':quitting, 'list':print_list}
+#main function
+if __name__ == "__main__":
+    while True:
+        send_thank_you_set = set(donors_list)
+        a = input('1: Send a Thank You \n2: Create a Report \n3: Send Letters to Everyone \n4: Quit \nChoose an Option: ')
+        if switch_func_dict.get(a, continue_func)() == False:
```

Comment:
to make this a little cleaner so that you don't have to add a bool return to each function just to do this check, change your `quitting` function:
```
def quitting():
    sys.exit()   # or quit()
```

Then you can get rid of all of the returns and remove the check for False. why is this important? Imagine you're on a team and you guys continue to add features to the mailroom - each time you add a feature, a dev has to remember to add the dummy return True to their functions.

Reply:
That makes sense. I imported sys, updated quitting function, and removed the True/False returns.

Reply:
I had to implement another "If" statement to send the user back to the main menu when they use 'quit' when asked for a name.

> def send_thank_you():
>     while True:
>         print('')
>         donor = input('Type a Full Name: ')
>         **if donor == 'quit':
>             return**
>         switch_func_dict.get(donor, continue_func)()
>

Reply:
hm shouldn't have to, it's a little hard to read without formatting (use triple back ticks for code formatting on github).

your main loop is a while True so if you end your function it will go back to the main menu on its own.

Reply:
```
def send_thank_you():
    while True:
        print('')
        donor = input('Type a Full Name: ')
        if donor == 'quit':
            return
        switch_func_dict.get(donor, continue_func)()
```
if the user types in 'quit', it still goes to this function and exits the program entirely (unless I keep the "if" statement). 
```
def quitting():
    sys.exit()
```

Reply:
do you have the latest version of your code you can link me? I can take a look at it - easier when I see the full code.

Reply:
Sure, I just pushed it.
https://github.com/josh1209/Self_Paced-Online/blob/master/students/Josh_HOff/lesson04/mailroom2.py

Reply:
I see now.

your `switch_func_dict` is for your main entry point. 

```
if response == 'quit':
    return
elif response == 'list':
    print_list()
else:
    # do add update/add donor stuff
```

I think you should keep the two flows separate: main entry point and send thank you flow

You could use switch dict in the send thank you flow as well but it is not required.